src/map/achievement.c

```diff
@@ -0,0 +1,1759 @@
+			return false;
+		}
+
+		obj->unique.weapon_lv = val;
+		obj->unique_type = CRITERIA_UNIQUE_WEAPON_LV;
+	} else if (achievement_criteria_weaponlv(type)) {
+		ShowError("achievement_readdb_validate_criteria_weaponlv: Criteria requires a WeaponType field. (Achievement: %d, Objective: %d). Skipping...\n", entry_id, obj_idx);
+		return false;
+	}
+
+	return true;
+}
+
+/**
+ * Validates achievement Id criteria of an objective while parsing an achievement entry.
+ * @param t        [in]     pointer to the config setting.
+ * @param obj      [in|out] pointer to the achievement objective entry being parsed.
+ * @param type     [in]     Type of the achievement being parsed.
+ * @param entry_id [in]     Id of the entry being parsed.
+ * @param obj_idx  [in]     Index of the objective entry being parsed.
+ * @return false on failure, true on success.
+ */
+bool achievement_readdb_validate_criteria_achievement(const struct config_setting_t *t, struct achievement_objective *obj, enum achievement_types type, int entry_id, int obj_idx)
+{
+	int val = 0;
+
+	nullpo_retr(false, t);
+	nullpo_retr(false, obj);
+
+	if (libconfig->setting_lookup_int(t, "Achieve", &val)) {
+		if (achievement->get(val) == NULL) {
+			ShowError("achievement_readdb_validate_criteria_achievement: Invalid Achievement %d provided as objective (Achievement %d, Objective %d). Skipping...\n", val, entry_id, obj_idx);
+			return false;
+		} else if (obj->unique_type != CRITERIA_UNIQUE_NONE) {
+			ShowError("achievement_readdb_validate_criteria_achievement: Unique criteria has already been set to type %d. (Achievement: %d, Objective: %d). Skipping...\n", (int) obj->unique_type, entry_id, obj_idx);
+			return false;
+		}
+
+		obj->unique.achieve_id = val;
+		obj->unique_type = CRITERIA_UNIQUE_ACHIEVE_ID;
+	} else if (type == ACH_ACHIEVE) {
+		ShowError("achievement_readdb_validate_criteria_achievement: Achievement type of ID %d requires an Achieve field in the objective criteria, setting not provided. Skipping...\n", entry_id);
+		return false;
+	}
+
+	return true;
+}
+
+/**
+ * Parses achievement objective entries of the current achievement db entry being parsed.
+ * @param conf       config pointer.
+ * @param entry      pointer to the achievement db entry being parsed.
+ * @param source     pointer to the string containing source/file name.
+ * @return false on failure, true on success.
+ */
+bool achievement_readdb_objectives(const struct config_setting_t *conf, struct achievement_data *entry)
+{
+	struct config_setting_t *t = NULL;
+
+	nullpo_retr(false, conf);
+	nullpo_retr(false, entry);
+
+	if ((t = libconfig->setting_get_member(conf, "Objectives")) && config_setting_is_group(t)) {
+		int i = 0;
+		// Initialize the buffer objective vector.
+		VECTOR_INIT(entry->objective);
+
+		for (i = 0; i < libconfig->setting_length(t) && i < MAX_ACHIEVEMENT_OBJECTIVES; i++) {
+			struct config_setting_t *tt = NULL;
```

Comment:
Reading separate objective should be in separate function (body of this for in separate function)